### ISSUE TITLE
Race Condition when used in combination with C2ME

### ISSUE THREAD

**Issue by iOmega8561** [nyuppo/MobVariants#33]
**Minecraft** 1.20.1 Fabric
**More Mob Variants** 1.3.0 and up
**C2ME**  0.2.0+alpha.11.5, default settings with *allowThreadedFeatures* set to false 

This issue started appearing after the 1.3.0 update and is still here with 1.3.0.1
Your mod is included in a custom pack, everything works fine until I start pre-generating the world (or just fly around generating chunks really). C2ME's stack trace mentions YungsApi but I can confirm that is not the cause of the race condition.
It is important to know that the server **does not** crash, it just keeps generating chunks and throwing errors.

Since there are more than 200 mods in this pack I also tested a blank fabric server with just More Mob Variants 1.3.0.1 and C2ME version 0.2.0+alpha.11.5 and the issue persists.

Of course if *threadedWorldGen* from C2ME is completely disabled, everything works fine.
Until the issue is resolved I will just roll back to version 1.2.2, which worked excellent for me.

```
[10:13:22] [C2ME worker #4/INFO]: [Chunky] Task running for minecraft:overworld. Processed: 8 chunks (0,05%), ETA: 0:04:07, Rate: 65,0 cps, Current: 2, 4
[10:13:23] [C2ME worker #3/INFO]: [Chunky] Task running for minecraft:overworld. Processed: 56 chunks (0,35%), ETA: 0:06:49, Rate: 39,2 cps, Current: 6, 1
[10:13:24] [C2ME worker #8/INFO]: [Chunky] Task running for minecraft:overworld. Processed: 115 chunks (0,71%), ETA: 0:05:46, Rate: 46,2 cps, Current: 13, 7
[10:13:24] [C2ME worker #6/ERROR]: ThreadLocalRandom accessed from a different thread (owner: Server thread, current: C2ME worker #6)
This is usually NOT a bug in C2ME, but a bug in another mod or in vanilla code. 
Possible solutions: 
  - Find possible causes in the stack trace below and 
    - if caused by another mod, report this to the corresponding mod authors 
    - if no other mods are involved, report this to C2ME 

java.util.ConcurrentModificationException: ThreadLocalRandom accessed from a different thread (owner: Server thread, current: C2ME worker #6)
	at com.ishland.c2me.fixes.worldgen.threading_issues.common.CheckedThreadLocalRandom.handleNotOwner(CheckedThreadLocalRandom.java:55) ~[c2me-fixes-worldgen-threading-is+alpha.11.5-9d70c476c16a4c1b.jar:?]
	at com.ishland.c2me.fixes.worldgen.threading_issues.common.CheckedThreadLocalRandom.isSafe(CheckedThreadLocalRandom.java:38) ~[c2me-fixes-worldgen-threading-is+alpha.11.5-9d70c476c16a4c1b.jar:?]
	at com.ishland.c2me.fixes.worldgen.threading_issues.common.CheckedThreadLocalRandom.method_43156(CheckedThreadLocalRandom.java:86) ~[c2me-fixes-worldgen-threading-is+alpha.11.5-9d70c476c16a4c1b.jar:?]
	at net.minecraft.class_6566.method_43055(class_6566.java:43) ~[server-intermediary.jar:?]
	at net.minecraft.class_1493.handler$dja000$moremobvariants$onReadCustomDataFromNbt(com/github/nyuppo/mixin/WolfVariantsMixin.java [moremobvariants.mixins.json]:47) ~[server-intermediary.jar:?]
	at net.minecraft.class_1308.method_5749(net/minecraft/class_1308.java:508) ~[server-intermediary.jar:?]
	at net.minecraft.class_1296.method_5749(net/minecraft/class_1296.java:113) ~[server-intermediary.jar:?]
	at net.minecraft.class_1429.method_5749(net/minecraft/class_1429.java:108) ~[server-intermediary.jar:?]
	at net.minecraft.class_1321.method_5749(net/minecraft/class_1321.java:52) ~[server-intermediary.jar:?]
	at net.minecraft.class_1493.method_5749(net/minecraft/class_1493.java:163) ~[server-intermediary.jar:?]
	at net.minecraft.class_1297.method_5651(net/minecraft/class_1297.java:1930) ~[server-intermediary.jar:?]
	at net.minecraft.class_1299.method_17839(net/minecraft/class_1299.java:549) ~[server-intermediary.jar:?]
	at net.minecraft.class_156.method_17974(net/minecraft/class_156.java:513) ~[server-intermediary.jar:?]
	at net.minecraft.class_1299.method_5892(net/minecraft/class_1299.java:548) ~[server-intermediary.jar:?]
	at net.minecraft.class_3499.yungsapi_getEntity(com/yungnickyoung/minecraft/yungsapi/mixin/EntityProcessorMixin.java [yungsapi_fabric.mixins.json]:126) ~[server-intermediary.jar:?]
	at net.minecraft.class_3499.handler$fco001$yungsapi$processEntities(com/yungnickyoung/minecraft/yungsapi/mixin/EntityProcessorMixin.java [yungsapi_fabric.mixins.json]:62) ~[server-intermediary.jar:?]
	at net.minecraft.class_3499.method_15172(net/minecraft/class_3499.java:347) ~[server-intermediary.jar:?]
	at net.minecraft.class_3781.method_16626(net/minecraft/class_3781.java:124) ~[server-intermediary.jar:?]
	at net.minecraft.class_3790.method_27236(net/minecraft/class_3790.java:89) ~[server-intermediary.jar:?]
	at net.minecraft.class_3790.method_14931(net/minecraft/class_3790.java:85) ~[server-intermediary.jar:?]
	at net.minecraft.class_3449.method_14974(net/minecraft/class_3449.java:100) ~[server-intermediary.jar:?]
	at net.minecraft.class_2794.method_38265(net/minecraft/class_2794.java:320) ~[server-intermediary.jar:?]
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422) ~[guava-31.1-jre.jar:?]
	at net.minecraft.class_2794.method_12102(net/minecraft/class_2794.java:319) ~[server-intermediary.jar:?]
	at net.minecraft.class_2806.method_51375(net/minecraft/class_2806.java:108) ~[server-intermediary.jar:?]
	at net.minecraft.class_2806$class_3768.doWork(class_2806.java:309) ~[server-intermediary.jar:?]
	at net.minecraft.class_2806.md644e30$c2me-threading-worldgen$lambda$runGenerationTask$0$5(com/ishland/c2me/threading/worldgen/mixin/MixinChunkStatus.java [c2me-threading-worldgen.mixins.json]:108) ~[server-intermediary.jar:?]
	at com.ishland.c2me.threading.worldgen.common.ChunkStatusUtils$ChunkStatusThreadingType$2.lambda$runTask$0(ChunkStatusUtils.java:111) ~[c2me-threading-worldgen-0.2.0+alpha.11.5-1d2f268f4f834365.jar:?]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
[10:13:24] [C2ME worker #6/INFO]: [STDOUT]: Negative index in crash report handler (0/23)
```

**Comment by Diaxium** [comment]
**ConcurrentModificationException** Fix in **CowVariantsMixin**.java

The ConcurrentModificationException related to ThreadLocalRandom occurs when multiple threads attempt to access and modify the same resource concurrently. The issue lies in the onReadCustomDataFromNbt method in CowVariantsMixin.java, where the synchronization on the ThreadLocalRandom class causes conflicts when multiple threads try to access it.

**Solution**
To resolve this issue, I've made an adjustment to remove the synchronization on ThreadLocalRandom, allowing each thread to access it independently without conflicts. This change fixes the concurrency issue and resolves the ConcurrentModificationException.

**Updated Code**
Here's the updated code:

```Java
@Override
    protected void onReadCustomDataFromNbt(NbtCompound nbt, CallbackInfo ci) {
        String variantKey = nbt.getString(MoreMobVariants.NBT_KEY);
        CowEntity self = (CowEntity)(Object)this;

        synchronized (ThreadLocalRandom.class) {
            if (!variantKey.isEmpty()) {
                Identifier variantId = variantKey.contains(":") ? new Identifier(variantKey) : MoreMobVariants.id(variantKey);
                variant = Variants.getVariant(EntityType.COW, variantId);
            } else {
                variant = Variants.getRandomVariant(EntityType.COW, self.getWorld().getRandom().nextLong(), self.getWorld().getBiome(self.getBlockPos()), null, self.getWorld().getMoonSize());
            }
        }

        // Update all players in the event that this is from modifying entity data with a command
        // This should be fine since the packet is so small anyways
        MinecraftServer server = ((Entity)(Object)this).getServer();
        if (server != null) {
            server.getPlayerManager().getPlayerList().forEach((player) -> {
                PacketByteBuf updateBuf = PacketByteBufs.create();
                updateBuf.writeInt(((Entity)(Object)this).getId());
                updateBuf.writeString(variant.getIdentifier().toString());

                ServerPlayNetworking.send(player, MMVNetworkingConstants.SERVER_RESPOND_BASIC_VARIANT_ID, updateBuf);
            });
        }
    }
```
By removing the synchronization, each thread can now access ThreadLocalRandom independently without conflicts, resolving the ConcurrentModificationException.

**Comment by nyuppo** [comment]
@Diaxium is it worth implementing this on the Forge version as well? I know the issue is only occuring when using C2ME, but I think the logic in question is also on the forge version. Just wondering if there are any potential drawbacks to be wary of.

**Comment by Diaxium** [comment]
@nyuppo Yes, it's worth implementing this fix on the Forge version as well, even if the issue is currently only occurring with the C2ME Fabric mod. Here's why:

## Benefits

1. **Consistency**: The logic is indeed present in both the Fabric and Forge versions. By applying the fix to both, you ensure consistency across platforms.
2. **Future-proofing**: Even if the issue isn't currently present in the Forge version, it's possible that a future mod or update could trigger the same concurrency problem.
3. **Thread safety**: The fix improves thread safety, which is essential for maintaining stability and preventing potential issues in multi-threaded environments.
4. **Best practice**: Applying the fix demonstrates a commitment to following best practices for concurrent programming, even if the issue isn't currently manifesting.

## Potential drawbacks

1. **Additional complexity**: The fix introduces a small amount of additional complexity, which might require additional testing and maintenance.
2. **Performance impact**: The addition of synchronization might have a negligible performance impact, potentially affecting very large-scale or high-concurrency environments. In the worst-case scenario, if many threads are frequently trying to access a synchronized block of code, this could lead to a significant performance degradation. This is because each thread has to wait for its turn to execute the synchronized block, leading to a lot of time spent waiting, especially in highly concurrent tasks. However, in less concurrent situations where the synchronized block is not accessed very frequently, the performance impact might be negligible.

Overall, the benefits of applying the fix to both Fabric and Forge versions outweigh the potential drawbacks. It's a proactive approach to ensuring the stability and thread safety of your mod.

## Note on Forge's case

It’s generally noted that If there is no actual need for synchronization (i.e., there is no risk of concurrent access conflicts), then using `synchronized` wouldn’t be necessary. However, in this specific case, implementing the fix on Forge is still worth considering given the benefits. 

Importantly, Minecraft is a multi-threaded game, and with the increasing popularity of performance mods that optimize threading and concurrency, the likelihood of concurrency issues arising in the future is higher.

**Comment by IntM4rco** [comment]
Can confirm that in Fabric 1.20.1

**Comment by fionera** [comment]
wow... This is the literal reason why chatGPT should be forbidden for coding

**Comment by nyuppo** [comment]
Removed a comment accusing me of using ChatGPT to write this mod lol

**Comment by fionera** [comment]
> Removed a comment accusing me of using ChatGPT to write this mod lol

I didn't meant you but Diaxium. Not only is the answer from their first comment wrong but also describes a completely different thing than what this actually is

**Comment by nyuppo** [comment]
> I didn't meant you but Diaxium. Not only is the answer from their first comment wrong but also describes a completely different thing than what this actually is

Ohh I see, thanks for the clarification! I haven't gotten around to actually implementing this yet, are you saying that Diaxium's comments aren't actually a good idea?

**Comment by Routhinator** [comment]
FWIW - I had this issue _without_ C2ME however the mod causing the LegacyRandomSource crash was unknown. C2ME adds a debugger and the log spam happens everytime an unsafe access method is used and spawns a trace to help diagnose it.

At any rate, removing MobVariants solves my LegacyRandomSource crash in 1.20.1 when pregenerating with Chunky, and I was able to find this thanks to C2ME.

**Comment by degiel1982** [comment]
Hi all I know I am late to discussion but here it happens to me as wel. I reverted back to 1.2.2 and it has no issues. I wonder what is the change in code that does this.

**Comment by Diamondback88** [comment]
Will this ever be fixed for Forge 1.20.1?

**Comment by sj-hub9796** [comment]
> Will this ever be fixed for Forge 1.20.1?

it already fixed but not released yet

**Comment by Diamondback88** [comment]
> > Will this ever be fixed for Forge 1.20.1?
> 
> it already fixed but not released yet

I don't think the mod author has been around for quite some time so idk if this will ever be published

**Comment by sj-hub9796** [comment]
> > > Will this ever be fixed for Forge 1.20.1?
> > 
> > 
> > it already fixed but not released yet
> 
> I don't think the mod author has been around for quite some time so idk if this will ever be published

um, you can clone this repo and build it self :P

**Comment by Diamondback88** [comment]
> > > > Will this ever be fixed for Forge 1.20.1?
> > > 
> > > 
> > > it already fixed but not released yet
> > 
> > I don't think the mod author has been around for quite some time so idk if this will ever be published
> 
> um, you can clone this repo and build it self :P

Yes but for putting it in public modpacks that won't work unless you got permission to republish the project
